x/data-proxy/keeper/msg_server.go

```diff
@@ -67,23 +70,118 @@ func (m msgServer) RegisterDataProxy(goCtx context.Context, msg *types.MsgRegist
 		return nil, types.ErrInvalidSignature
 	}
 
-	err = m.DataProxyConfigs.Set(ctx, pubKeyBytes, types.ProxyConfig{
+	proxyConfig := types.ProxyConfig{
 		PayoutAddress: msg.PayoutAddress,
 		Fee:           msg.Fee,
 		Memo:          msg.Memo,
 		FeeUpdate:     nil,
 		AdminAddress:  msg.AdminAddress,
-	})
+	}
+
+	err = proxyConfig.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.DataProxyConfigs.Set(ctx, pubKeyBytes, proxyConfig)
 	if err != nil {
 		return nil, err
 	}
 
 	return &types.MsgRegisterDataProxyResponse{}, nil
 }
 
-func (m msgServer) EditDataProxy(_ context.Context, _ *types.MsgEditDataProxy) (*types.MsgEditDataProxyResponse, error) {
-	// TODO
-	return &types.MsgEditDataProxyResponse{}, nil
+func (m msgServer) EditDataProxy(goCtx context.Context, msg *types.MsgEditDataProxy) (*types.MsgEditDataProxyResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if err := msg.Validate(); err != nil {
+		return nil, err
+	}
+
+	pubKeyBytes, err := hex.DecodeString(msg.PubKey)
+	if err != nil {
+		return nil, types.ErrInvalidHex.Wrap(err.Error())
+	}
+
+	proxyConfig, err := m.DataProxyConfigs.Get(ctx, pubKeyBytes)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return nil, types.ErrUnknownDataProxy.Wrapf("no data proxy registered for %s", msg.PubKey)
+		}
+		return nil, err
+	}
+
+	if msg.Sender != proxyConfig.AdminAddress {
+		return nil, types.ErrUnauthorized
+	}
+
+	err = proxyConfig.UpdateBasic(msg.NewPayoutAddress, msg.NewMemo)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there is no new fee we can terminate early
+	if msg.NewFee == nil {
+		err = m.DataProxyConfigs.Set(ctx, pubKeyBytes, proxyConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		return &types.MsgEditDataProxyResponse{}, nil
+	}
+
+	return m.ProcessProxyFeeUpdate(ctx, pubKeyBytes, &proxyConfig, msg)
+}
+
+func (m msgServer) ProcessProxyFeeUpdate(ctx sdk.Context, pubKeyBytes []byte, proxyConfig *types.ProxyConfig, msg *types.MsgEditDataProxy) (*types.MsgEditDataProxyResponse, error) {
+	params, err := m.Keeper.Params.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	updateDelay := params.MinFeeUpdateDelay
+	// Validate custom delay if passed
+	if msg.FeeUpdateDelay != types.UseMinimumDelay {
+		if msg.FeeUpdateDelay < params.MinFeeUpdateDelay {
+			return nil, types.ErrInvalidDelay.Wrapf("minimum delay %d, got %d", params.MinFeeUpdateDelay, msg.FeeUpdateDelay)
+		}
+
+		updateDelay = msg.FeeUpdateDelay
+	}
+
+	// Determine update height
+	updateHeight := ctx.BlockHeight() + int64(updateDelay)
+	feeUpdate := &types.FeeUpdate{
+		NewFee:       *msg.NewFee,
+		UpdateHeight: updateHeight,
+	}
+
+	// Retain previous pending update so it can be removed
+	previousPendingUpdate := proxyConfig.FeeUpdate
```

Comment:
Why don't we just check the value of `proxyConfig.FeeUpdate` and remove if applicable here?

Reply:
Great question, no real reason.